### ISSUE TITLE
[REQ] Support for code licensing

### ISSUE THREAD

**Issue by dkarlovi** [OpenAPITools/openapi-generator#7535]
### Is your feature request related to a problem? Please describe.

[OpenAPI v3 supports supplying licensing information](https://swagger.io/docs/specification/api-general-info/).

## Describe the solution you'd like

It would make sense to propagate this information into the generated code, for example creating a `LICENSE` file in the root of the package, also allowing the generators access to this information, for example in `php` you'd add the license information to `composer.json`.

**Comment by dkarlovi** [comment]
@wing328 how do you feel about this one?

**Comment by jimschubert** [comment]
We already pass `license` to templates. It's up to specific templates to incorporate this. 

The license in an OpenAPI document is the API's or Spec's license, not necessarily the license of the generated code. We can't assume such. We do, however, allow for easily extending templates so users can apply something like `{{#license}}{{license}}{{/license}}{{^license}}UKNONWN{{/license}}` if required.

I think it's best to leave it up to users to define the `LICENSE` file in their output directory, as that is above and beyond the responsibility of our tool.

**Comment by dkarlovi** [comment]
@jimschubert arguably, the license is one of the most important aspects of the code being generated, [the tool goes as far a claiming authorship (basically, ownership) of the code](https://github.com/flexolabs/reddit-client/blob/25feb73a9df32154d14598c298726c0cb3d017bb/src/ApiException.php#L14). The generator user should be able to apply one, there's at least two things it can do:

1. generate the correct `LICENSE` file directly in the root, regardless of the generator
2. pass the actual license to the generator, which can then embedded it where it needs it (docs, `composer.json`/`package.json` equivalent)

Maybe adding a `--license` flag to the generate command, defaulting to the license defined in the spec?

The ownership / licensing of the generated code should be made more transparent, this would be a good first step toward it.

**Comment by jimschubert** [comment]
@dkarlovi Again, we already do option 2. It is up to generators to consume the license details, or up to users to extend points where licenses should be applied. I don't understand your reasoning that having an author tag in code authored by our tool/community suggests that we should also output licensing details... those are totally different concepts and not related in any way.

As far as outputting a `LICENSE` file... how many licenses do you propose we maintain the full license text? Only the popular ones, or do we persist the hundreds of open source license texts? How does maintaining all that license text (licensed CC-0) affect the license of our tool (Apache 2.0)? What license do we output when we don't have the text? What happens, legally, if we choose the wrong license and a user doesn't notice before they release? How do we lookup the license text? Who maintains it?

We previously introduced license outputs in generated code and received backlash from the community. The feedback was essentially that we _cannot_ make any assumptions or mistakes about licenses. So, we removed license output and left it up to users.

As an example of the efforts we've made, most of our generators support a `partial_header` template which users can define locally and apply to source outputs. Recently, I've added the ability for users to pass their own template files to a generator and control where it outputs. Not only are these options trivial for users, they remove the need for adding logical, legal, and maintenance complexity to our tool (my questions above).

Do you think it would suffice to add a section to our docsite explaining how to apply a license to generated code?

**Comment by dkarlovi** [comment]
You pass the license from the spec, but you've said
> The license in an OpenAPI document is the API's or Spec's license, not necessarily the license of the generated code. We can't assume such. 

So which is it?

> As far as outputting a LICENSE file... how many licenses do you propose we maintain the full license text?

You could always use the [OSI license list](https://opensource.org/licenses/alphabetical) as a starting point, that would likely cover 99% of the cases.

So, the workflows:

## Explicit `--license`
1. I run `generate --license MIT --copyright SomeOrg Inc.`
2. general generator pulls https://opensource.org/licenses/MIT into a text file called `LICENSE`, pass it the copyright information
3. replaces the `LICENSE` passed to the generator with this one

## Implicit, from spec
1. I run `generate --copyright SomeOrg Inc.`
2. generator finds `AGPL-3.0` license in spec
3. general generator pulls https://opensource.org/licenses/AGPL-3.0 into a text file called `LICENSE`, pass it the copyright information
4. passes that one to the generator, which it already does

## Implicit, nothing in spec

Same as currently.

**Comment by wing328** [comment]
> Maybe adding a --license flag to the generate command, defaulting to the license defined in the spec?

As @jimschubert has mentioned, we did that previously but based on feedback from our users, it's not a good idea. 

The auto-generated code is unlicensed by default: https://github.com/OpenAPITools/openapi-generator#34---license-information-on-generated-code, which seems to be a proper middle ground for most of our users so far.

> You could always use the OSI license list as a starting point, that would likely cover 99% of the cases.

Based on the feedback from our users before, OSI license may or may not meet their requirement as it may require lengthy review from their legal departments.

 > general generator pulls https://opensource.org/licenses/MIT into a text file called LICENSE, pass it the copyright information

Previously I recommend a user download the license file using `wget` as part of a script that includes using openapi-generator to generate the code and he's happy with the approach.

(I can see your good intention to try to automate more for the users which is something we tried a while ago. Thanks again for the suggestion and bringing it up to the core team's attention.)
